x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,214 @@
+)
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func init() {
+	// gather all req names to a list
+	allReqNames = append(allReqNames, stakeReqName)
+
+	// init strategy
+	uniformStrategy = make(ScoreStrategy)
+	for _, reqName := range allReqNames {
+		uniformStrategy[reqName] = 1
+	}
+
+	if len(allReqNames) != len(uniformStrategy) {
+		panic("strategy does not contain all score reqs")
+	}
+}
+
+// get the overall requirements from the policy and assign slots that'll fulfil them
+// TODO: this function should be changed in the future since it only supports stake reqs
+func CalcSlots(policy planstypes.Policy, minStake sdk.Int) []*PairingSlot {
+	// init slot array (should be as the number of providers to pair)
+	slots := make([]*PairingSlot, policy.MaxProvidersToPair)
+
+	// all slots should consider the stake, so we init them with stakeReq
+	stakeReq := StakeReq{MinStake: minStake}
+	slotReqs := map[string]ScoreReq{stakeReq.GetName(): stakeReq}
+	for i := range slots {
+		slots[i] = NewPairingSlot()
+		slots[i].Reqs = slotReqs
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []*PairingSlot {
+	uniqueSlots := make(map[string]*PairingSlot)
+	if len(slots) == 0 {
+		utils.LavaFormatError("no slots", sdkerrors.ErrLogic)
+		return []*PairingSlot{}
+	}
+
+	for i := 0; i < len(slots); i++ {
+		slot := slots[i]
+		key := slot.GetSlotKey()
+
+		if existingSlot, ok := uniqueSlots[key]; ok {
+			// Duplicate slot found
+			existingSlot.Count++
+		} else {
+			// New unique slot
+			uniqueSlots[key] = slot
+		}
+	}
+
+	// Convert map values to slice
+	uniqueSlotList := make([]*PairingSlot, 0, len(uniqueSlots))
+	for _, slot := range uniqueSlots {
+		uniqueSlotList = append(uniqueSlotList, slot)
+	}
+
+	return uniqueSlotList
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// CalcPairingScore calculates the final pairing score for a pairing slot (with strategy)
+// For efficiency purposes, we calculate the score on a diff slot which represents the diff reqs of the current slot
+// and the previous slot
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			newScoreComp = uint64(math.Pow(float64(newScoreComp), float64(weight)))
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				if prevReqScoreComp == 0 {
+					utils.LavaFormatFatal("previous req score is zero", fmt.Errorf("invalid req score"),
```

Comment:
replace with: `utils.LavaFormatError("critical: ...", ...)`
(the `LavAFormatFatal()` version would exit with an error)